source/particle/world.cc

```diff
@@ -617,6 +614,158 @@ namespace aspect
     {
+
+      /**
+       * Wrapper around dealii::FEPointEvaluation to choose number of components dynamically.
+       */
+      template <int dim>
+      class DynamicFEPointEvaluation
+      {
+        public:
+          DynamicFEPointEvaluation(const unsigned int first_component, const unsigned int n_components)
+            : first_component (first_component),
+              n_components(n_components)
+          {}
+
+          virtual ~DynamicFEPointEvaluation() = default;
+
+          int first_component;
+          int n_components;
+
+          virtual void evaluate(const ArrayView<double> &solution_values,
+                                const EvaluationFlags::EvaluationFlags flags) = 0;
+
+          virtual
+          std::vector<Tensor<1,dim>>
+          get_gradient(const unsigned int evaluation_point) const = 0;
+
+          virtual
+          std::vector<double>
+          get_value(const unsigned int evaluation_point) const = 0;
+
+      };
+
+
+
+      template <int dim>
+      Tensor<1,dim> to_tensor(const Tensor<1,dim> &in)
+      {
+        return in;
+      }
+
+
+
+      template <int dim>
+      Tensor<1,1> to_tensor(const double in)
+      {
+        Tensor<1,1> result;
+        result[0] = in;
+        return result;
+      }
+
+
+
+      template <int dim, int n_components>
+      Tensor<1,n_components,Tensor<1,dim>> to_tensor2(const Tensor<1,n_components,Tensor<1,dim>> &in)
+      {
+        return in;
+      }
+
+
+
+      template <int dim, int n_components>
+      Tensor<1,1,Tensor<1,dim>> to_tensor2(const Tensor<1,dim> &in)
+      {
+        Tensor<1,1,Tensor<1,dim>> result;
+        result[0] = in;
+        return result;
+      }
+
+
+
+      template <int dim, int n_components>
+      class DynamicFEPointEvaluationImpl: public DynamicFEPointEvaluation<dim>
+      {
+        public:
+          DynamicFEPointEvaluationImpl(NonMatching::MappingInfo<dim> &mapping,
+                                       const FiniteElement<dim> &fe,
+                                       const unsigned int        first_selected_component)
+            : DynamicFEPointEvaluation<dim>(first_selected_component, n_components),
+              evaluation(mapping, fe, first_selected_component)
+          {}
+
+          void evaluate(const ArrayView<double> &solution_values,
+                        const EvaluationFlags::EvaluationFlags flags) override
+          {
+            evaluation.evaluate(solution_values, flags);
+          }
+
+          virtual
+          std::vector<Tensor<1,dim>>
+          get_gradient(const unsigned int evaluation_point) const override
+          {
+            const Tensor<1,n_components,Tensor<1,dim>> x = to_tensor2<dim,n_components>(evaluation.get_gradient(evaluation_point));
+            std::vector<Tensor<1,dim>> result (n_components);
+            for (int c=0; c<n_components; ++c)
+              result[c] = x[c];
+            return result;
+          }
+
+          virtual
```

Comment:
no need for `virtual` if you already specify `override` here and for the functions above.